### DISCUSSION TITLE
[Dove] Service Declarations In Typescript

### DISCUSSION THREAD

**Discussion by forgot** [feathersjs/feathers#2359] [Q&A] [answered]
Before Dove, when creating a service in Typescript with the CLI, the output is like this:
```typescript
// Initializes the `Messages` service on path `/messages`
import { ServiceAddons } from '@feathersjs/feathers';
import { Application } from '../../declarations';
import { Messages } from './messages.class';
import createModel from '../../models/messages.model';
import hooks from './messages.hooks';

// Add this service to the service type index
declare module '../../declarations' {
  interface ServiceTypes {
    'messages':  Messages & ServiceAddons<any>;
  }
}

export default function (app: Application): void {
  const options = {
    Model: createModel(app),
    paginate: app.get('paginate')
  };

  // Initialize our service with any options it requires
  app.use('/messages', new Messages(options, app));

  // Get our initialized service so that we can register hooks
  const service = app.service('messages');

  service.hooks(hooks);
}
```
In Dove, Typescript throws an error at `app.use('/messages', new Messages(options, app));`:
```
No overload matches this call.
  The last overload gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'RequestHandler<ParamsDictionary, any, any, ParsedQs, Record<string, any>>'.
```

Removing `ServiceAddons<any>` and the `/` from `/messages` clears everything up, but not allowing a `/` in the service path keeps me from being able to do things like create namespaced services.

Removing the above mentioned, the following service declaration doesn't cause the Typescript compiler to complain:
```typescript
// Initializes the `messages` service on path `/messages`
import { Application } from '../../declarations';
import { Messages } from './messages.class';
import createModel from '../../models/messages.model';
import hooks from './messages.hooks';

// Add this service to the service type index
declare module '../../declarations' {
  interface ServiceTypes {
    'messages': Messages
  }
}

export default function (app: Application): void {
  const options = {
    Model: createModel(app),
    paginate: app.get('paginate')
  };

  // Initialize our service with any options it requires
  app.use('messages', new Messages(options, app));

  // Get our initialized service so that we can register hooks
  const service = app.service('messages');

  service.hooks(hooks);
}
```

My question is, what is the proper way to structure this service declaration in Dove and allow for `/`'s in the path?

**Comment by daffl** [comment, marked as answer]
I don't think there is a way for allowing `/` while having a service declaration interface unless you explicitly add it. I don't believe TypeScript has property pattern/regex matching abilities.

**Reply by daffl** [reply]
In reality, any leading or tailing `/` always gets stripped out and never really indicated anything useful.

**Reply by forgot** [reply]
Understood. What would be the best way to create an internal service? I know I can disallow external providers, but is there any way to namespace the service?

**Reply by daffl** [reply]
There's a couple of options, one would be creating a new `app` that is only used internally only. In v5 you can also do

```js
app.use('messages', new Messages(options, app), {
  methods: []
});
```

Which will make no methods ever available externally at all.

**Reply by forgot** [reply]
That looks great! Thank you!
